"No matching call to `ebo_get`" regression

Consider the following code:

```cpp
#include <boost/hana.hpp>

namespace bh = boost::hana;

int main()
{
    auto m = bh::make_map();
    bh::int_<0> key;
    bh::int_<1> x;
    auto new_map =
        bh::insert(m, bh::make_pair(key, bh::make_pair(x, bh::false_c)));
}
```

It compiles and works properly with the last version of Boost. [**on wandbox**](https://wandbox.org/permlink/F80HpydVhAwQt5hX).

With the `master` branch of this repository, the code fails to compile:

> 
> In file included from /usr/local/include/boost/hana/detail/struct_macros.hpp:27:0,
>                  from /usr/local/include/boost/hana/adapt_adt.hpp:15,
>                  from /usr/local/include/boost/hana.hpp:59,
>                  from hanabug.cpp:1:
> /usr/local/include/boost/hana/pair.hpp: In instantiation of ‘static constexpr decltype(auto) boost::hana::first_impl<boost::hana::pair_tag>::apply(P&&) [with P = boost::hana::pair<boost::hana::integral_constant<int, 0>, boost::hana::pair<boost::hana::integral_constant<int, 1>, boost::hana::integral_constant<bool, false> > >&]’:
> /usr/local/include/boost/hana/first.hpp:34:28:   required from ‘constexpr decltype(auto) boost::hana::first_t::operator()(Pair&&) const [with Pair = boost::hana::pair<boost::hana::integral_constant<int, 0>, boost::hana::pair<boost::hana::integral_constant<int, 1>, boost::hana::integral_constant<bool, false> > >&]’
> /usr/local/include/boost/hana/map.hpp:307:45:   required from ‘static constexpr auto boost::hana::insert_impl<boost::hana::map_tag>::apply(Map&&, Pair&&) [with Map = boost::hana::detail::map_impl<boost::hana::detail::hash_table<>, boost::hana::basic_tuple<> >&; Pair = boost::hana::pair<boost::hana::integral_constant<int, 0>, boost::hana::pair<boost::hana::integral_constant<int, 1>, boost::hana::integral_constant<bool, false> > >]’
> /usr/local/include/boost/hana/insert.hpp:28:68:   required from ‘constexpr decltype(auto) boost::hana::insert_t::operator()(Set&&, Args&& ...) const [with Set = boost::hana::detail::map_impl<boost::hana::detail::hash_table<>, boost::hana::basic_tuple<> >&; Args = {boost::hana::pair<boost::hana::integral_constant<int, 0>, boost::hana::pair<boost::hana::integral_constant<int, 1>, boost::hana::integral_constant<bool, false> > >}]’
> hanabug.cpp:10:83:   required from here
> /usr/local/include/boost/hana/pair.hpp:166:49: error: no matching function for call to ‘ebo_get<boost::hana::detail::pix<0> >(boost::hana::pair<boost::hana::integral_constant<int, 0>, boost::hana::pair<boost::hana::integral_constant<int, 1>, boost::hana::integral_constant<bool, false> > >&)’
>          { return detail::ebo_get<detail::pix<0>>(static_cast<P&&>(p)); }
>                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
> In file included from /usr/local/include/boost/hana/pair.hpp:17:0,
>                  from /usr/local/include/boost/hana/detail/struct_macros.hpp:27,
>                  from /usr/local/include/boost/hana/adapt_adt.hpp:15,
>                  from /usr/local/include/boost/hana.hpp:59,
>                  from hanabug.cpp:1:
> 

---

Related ECST issue [#30](https://github.com/SuperV1234/ecst/issues/30).

With clang 3.9, I get a different error message:
```
/usr/local/src/hana/include/boost/hana/pair.hpp:93:50: error: cannot cast from lvalue of type
      'boost::hana::integral_constant<int, 1>' to rvalue reference type
      'boost::hana::integral_constant<int, 0> &&'; types are not compatible
            : detail::ebo<detail::pix<0>, First>(static_cast<T&&>(detail::ebo_get<detail::pix<0>>(other)))
                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

It looks like the call to `ebo_get` is is trying to forward a returned value.

--- a/include/boost/hana/pair.hpp
+++ b/include/boost/hana/pair.hpp
@@ -161,16 +161,46 @@ BOOST_HANA_NAMESPACE_BEGIN
 
     template <>
     struct first_impl<pair_tag> {
-        template <typename P>
-        static constexpr decltype(auto) apply(P&& p)
-        { return detail::ebo_get<detail::pix<0>>(static_cast<P&&>(p)); }
+        template <typename First, typename Second>
+        static constexpr decltype(auto) apply(hana::pair<First, Second>& p) {
+            return detail::ebo_get<detail::pix<0>>(
+                static_cast<detail::ebo<detail::pix<0>, First>&>(p)
+            );
+        }
+        template <typename First, typename Second>
+        static constexpr decltype(auto) apply(hana::pair<First, Second> const& p) {
+            return detail::ebo_get<detail::pix<0>>(
+                static_cast<detail::ebo<detail::pix<0>, First> const&>(p)
+            );
+        }
+        template <typename First, typename Second>
+        static constexpr decltype(auto) apply(hana::pair<First, Second>&& p) {
+            return detail::ebo_get<detail::pix<0>>(
+                static_cast<detail::ebo<detail::pix<0>, First>&&>(p)
+            );
+        }
     };
 
     template <>
     struct second_impl<pair_tag> {
-        template <typename P>
-        static constexpr decltype(auto) apply(P&& p)
-        { return detail::ebo_get<detail::pix<1>>(static_cast<P&&>(p)); }
+        template <typename First, typename Second>
+        static constexpr decltype(auto) apply(hana::pair<First, Second>& p) {
+            return detail::ebo_get<detail::pix<1>>(
+                static_cast<detail::ebo<detail::pix<1>, Second>&>(p)
+            );
+        }
+        template <typename First, typename Second>
+        static constexpr decltype(auto) apply(hana::pair<First, Second> const& p) {
+            return detail::ebo_get<detail::pix<1>>(
+                static_cast<detail::ebo<detail::pix<1>, Second> const&>(p)
+            );
+        }
+        template <typename First, typename Second>
+        static constexpr decltype(auto) apply(hana::pair<First, Second>&& p) {
+            return detail::ebo_get<detail::pix<1>>(
+                static_cast<detail::ebo<detail::pix<1>, Second>&&>(p)
+            );
+        }
     };
 BOOST_HANA_NAMESPACE_END
 

--- a/test/issues/github_331.cpp
+++ b/test/issues/github_331.cpp
@@ -1,0 +1,45 @@
+// Copyright Louis Dionne 2013-2017
+// Distributed under the Boost Software License, Version 1.0.
+// (See accompanying file LICENSE.md or copy at http://boost.org/LICENSE_1_0.txt)
+
+#include <boost/hana/at.hpp>
+#include <boost/hana/first.hpp>
+#include <boost/hana/integral_constant.hpp>
+#include <boost/hana/pair.hpp>
+#include <boost/hana/second.hpp>
+#include <boost/hana/tuple.hpp>
+
+#include <type_traits>
+namespace hana = boost::hana;
+
+
+// In GitHub issue #331, we noticed that `first` and `second` could sometimes
+// return the wrong member in case of nested pairs. This is due to the way we
+// inherit from base classes to enable EBO. We also check for `basic_tuple`,
+// because both are implemented similarly.
+
+int main() {
+    {
+        using Nested = hana::pair<hana::int_<1>, hana::int_<2>>;
+        using Pair = hana::pair<hana::int_<0>, Nested>;
+        Pair pair{};
+
+        auto a = hana::first(pair);
+        static_assert(std::is_same<decltype(a), hana::int_<0>>{}, "");
+
+        auto b = hana::second(pair);
+        static_assert(std::is_same<decltype(b), Nested>{}, "");
+    }
+
+    {
+        using Nested = hana::basic_tuple<hana::int_<1>, hana::int_<2>>;
+        using Tuple = hana::basic_tuple<hana::int_<0>, Nested>;
+        Tuple tuple{};
+
+        auto a = hana::at_c<0>(tuple);
+        static_assert(std::is_same<decltype(a), hana::int_<0>>{}, "");
+
+        auto b = hana::at_c<1>(tuple);
+        static_assert(std::is_same<decltype(b), Nested>{}, "");
+    }
+}

